nni/algorithms/compression/pytorch/quantization/quantizers.py

```diff
@@ -545,4 +545,81 @@ def export_model(self, model_path, calibration_path=None, onnx_path=None, input_
+        for layer, config in modules_to_compress:
+            layer.module.register_parameter("zero_point", torch.nn.Parameter(torch.Tensor([0.0])))
+            layer.module.register_parameter("scale", torch.nn.Parameter(torch.Tensor([1.0])))
+            if "weight" in config.get("quant_types", []):
+                # todo: support per-channel quantization for weight since TensorRT it for conv weight
+                q_bit = get_bits_length(config, "weight")
+                qmax = 2 ** (q_bit - 1) - 1
+                qmin = -2 ** (q_bit - 1)
+                init_weight_scale = layer.module.weight.data.detach().abs().mean() * 2 / (qmax ** 0.5)
+                layer.module.scale = torch.nn.Parameter(init_weight_scale)
+                layer.module.weight_qmax = qmax
+                layer.module.weight_qmin = qmin
+
+            # todo: in the origin paper, the initial value of activation is calculated from first input batch
+            if "output" in config.get("quant_types", []):
+                q_bit = get_bits_length(config, "")
+                qmax = 2 ** (q_bit - 1) - 1
+                qmin = -2 ** (q_bit - 1)
+                layer.module.activation_qmax = qmax
+                layer.module.activation_qmin = qmin
+            # add zero_point and scale to optimizer since they are updated through the gradient
+            self.optimizer.add_param_group({"params": layer.module.zero_point})
+            self.optimizer.add_param_group({"params": layer.module.scale})
+
+    @staticmethod
+    def grad_scale(x, scale):
+        """
+            Used to scale the gradient
+        """
+        y = x
+        y_grad = x * scale
+        return (y - y_grad).detach() + y_grad
+
+    @staticmethod
+    def round_pass(x):
+        """
+            A simple way to execute `round` operation with grad set to 1
+        """
+        y = x.round()
+        y_grad = x
+        return (y - y_grad).detach() + y_grad
+
+    def quantize(self, x, scale, zero_point, qmin, qmax):
+        grad_scale_factor = 1.0 / ((qmax * x.numel()) ** 0.5)
+        scale = self.grad_scale(scale, grad_scale_factor)
+        x = x / scale + zero_point
+        x = torch.clamp(x, qmin, qmax)
+        x = self.round_pass(x)
+        x = (x - zero_point) * scale
+        return x
+
+    def quantize_weight(self, wrapper, **kwargs):
+        module = wrapper.module
+
+        # todo: add support for quantize bias. If we use TensorRT as backend, there is no need to quantize
+        # bias
+        old_weight = module.old_weight
+        weight = self.quantize(old_weight, module.scale, module.zero_point, module.weight_qmin, module.weight_qmax)
+        module.weight = weight
+        return weight
+
+    def quantize_output(self, output, wrapper, **kwargs):
+        module = wrapper.module
+        output = self.quantize(output, module.scale, module.zero_point, module.activation_qmin, module.activation_qmax)
+        return output
```

Comment:
Can this quantization algorithm support exporting model and related quantization parameters? If yes, maybe we can consider adding function `export_model()` based on what parameters should export to inference framework like TensorRT.

Reply:
I will check it out